### ISSUE TITLE
Make pyproject.toml + uv.lock the single dependency owner, and retire environment.yml

### ISSUE THREAD

**Issue by derrring** [derrring/MFGArchon#2167]
Two hand-maintained manifests declare the same dependencies. #2164 measured the fork: eight packages
the library imports were missing from `environment.yml`, six that #1687 had already removed were
still there, and six version floors disagree — `environment.yml` looser in every case, so a conda
environment can satisfy it with `osqp 0.6` while `pyproject.toml` requires `>=1.0`.

#2166 reconciles them. That patches the fork; it does not close it. This issue closes it.

## Target

`pyproject.toml` owns **what the project depends on**. `uv.lock` owns **which exact versions**.
`environment.yml` is deleted.

## The objection this has to answer, and the measured answer

Retiring conda was thought to cost BLAS flexibility, because conda-forge is the only ecosystem where
the implementation is swappable after install (`libblas=*=*mkl|openblas|accelerate|blis|netlib`).
PyPI wheels bake theirs in, and there is no maintained MKL-linked numpy on PyPI: `numpy-mkl` is
1.10.2 sdist-only, `intel-numpy` stopped at 1.21.5, `scipy-mkl` does not exist, and the `mkl` package
is the Intel library itself, not a numpy build.

**That flexibility does not require a second manifest.** BLAS is a property of how numpy/scipy were
built, not of what the project depends on. Measured: `uv pip install` treats a conda-installed numpy
as satisfying and leaves it alone, including under `--reinstall-package`.

```
conda create -n mfg-mkl python=3.12 "libblas=*=*mkl" numpy scipy   # substrate, ~30 packages
uv pip install -e ".[all,dev]" --python <that env>                 # dependencies from pyproject
```

So what replaces `environment.yml` is two lines of documentation, not a manifest. This also *gains*
something on macOS: the PyPI numpy wheel links Apple Accelerate here, while the current conda
environment links the generic `blas 3.9.0`.

## Phase 1

1. **Extras vs groups.** Dev tooling (pytest, ruff, mypy, pre-commit, sphinx) moves to
   `[dependency-groups]` — not published metadata. Compute backends stay in
   `[project.optional-dependencies]`, because a user installing the package must be able to ask for
   them. Getting this backwards makes the backends uninstallable.
2. **Lock scope.** `[tool.uv] environments = ["sys_platform == 'darwin'"]` to start. Measured cost of
   the universal default: a torch-bearing lock goes from 11 packages to 30, of which 15 are
   `nvidia-*`, all linux-gated. Adding linux later is one line plus `uv lock`, so this defers the
   platform rather than excluding it.
3. **Backend builds.** Verified working: `[project.optional-dependencies]` `cpu`/`cuda` extras,
   `[[tool.uv.index]]` for `https://download.pytorch.org/whl/cpu`, `[tool.uv.sources]` selecting by
   extra, and `[tool.uv.conflicts]` declaring them mutually exclusive. One lock then carries both
   `torch 2.13.0` (PyPI, CUDA) and `torch 2.13.0+cpu`.
4. **Name mapping is nearly empty.** 37 of 39 conda names resolve on PyPI unchanged. Drop `python`
   (uv manages it via `requires-python`), `pip` and `conda-build`; rename `pytorch` → `torch`, since
   PyPI `pytorch` is a 1.0.2 placeholder.
5. **Record the BLAS in the warning census.** #2158 added a toolchain block to the census;
   `np.show_config(mode="dicts")["Build Dependencies"]["blas"]["name"]` belongs in it. Moving to uv
   changes the BLAS on macOS, and without this the resulting identity drift is unattributable — the
   exact failure #2158 exists to prevent.
6. Delete `environment.yml`, drop direction two of `scripts/check_manifests.py`, and document the
   conda substrate recipe.

## Not in scope here

`uv.lock`'s current state is #2138 (tracked, stale, unsatisfied, read by nothing) and the ruff
disagreement is #2147. Both must land before or with step 2, since this issue makes the lock
authoritative.

## Open decisions

- Whether to keep macOS-only locking past Phase 1, or declare linux as soon as anyone needs it.
- Whether MKL is worth pursuing at all. The project is mixed sparse/small-dense — 102 `scipy.sparse`
  sites, 49 sparse factorisations, against 36 dense `np.linalg` calls concentrated in GFDM stencil
  construction and QP utilities. That is the regime where BLAS differences are usually small.
  Nobody has measured it, and step 5 is the cheap prerequisite for anyone who wants to.

**Comment by derrring** [comment]
**Step 1 landed in #2171.** `dev` and `docs` are now `[dependency-groups]`; `core`, `nn`, `all` and
`numerical` stay extras. Fourteen call sites moved to `--group dev`, and
`tests/unit/test_install_commands_name_real_extras_2167.py` now fails on any install command naming
an extra or group `pyproject.toml` does not declare.

Two things measured along the way that this issue should decide, both out of step 1's scope:

- **`core` is a pure duplicate of `[project.dependencies]`.** All six entries are byte-identical
  strings already declared there, so `pip install mfgarchon[core]` installs exactly what
  `pip install mfgarchon` does. Six version floors declared twice, with nothing keeping them in
  step. The pyproject comment calls it the "minimal (CI, Docker, embedding)" tier, which it has not
  been since the default gained the same packages. Either it means something different from the
  default and should say what, or it is dead weight.
- **`numerical` duplicates `cvxpy>=1.4` from `all`.** One floor declared twice. Unlike `core` it has
  live user-facing consumers — `mfgarchon/utils/dependencies.py` prints
  `pip install mfgarchon[numerical]` — so removing it is a decision about the install surface, not
  just a cleanup. #2170 covers the neighbouring problem of extras that are *named but never
  declared*, and the two should be decided together.

Step 5 (record the BLAS) is #2168, stacked on #2163. Steps 2/3/4/6 remain, and #2138's lock
retirement (#2169) was the prerequisite for step 2.

**Comment by derrring** [comment]
**Step 5 reworked after adversarial review (#2168), and one piece was deliberately dropped.**

A runtime BLAS reading via `threadpoolctl.threadpool_info()` was designed, measured working, and
then removed from the PR. It does name what conda hides — measured `openblas 0.3.33` at
`lib/libopenblas.0.dylib` in `mfg_env`, where `numpy.show_config` says only `blas` — so the
capability is real. Three measurements decided against shipping it now:

- **Null in the environment being left and in the one being moved to.** Nothing in `pyproject.toml`
  or `environment.yml` installs `threadpoolctl`, and it is absent from a fresh PyPI venv too. The
  value quoted in the first version of that PR was obtainable only by putting the module on
  `PYTHONPATH` by hand.
- **Silent under Accelerate regardless.** threadpoolctl registers controllers for openblas, blis,
  flexiblas, mkl and openmp; Accelerate is a system framework it does not hook. So in the
  destination environment it answers nothing even when installed.
- **Its thread count churns.** `OMP_NUM_THREADS=1` and `=4` produce different values;
  `.uvrc:14` sets 4 and `ci.yml:218` sets 1. That is drift on an axis nobody changed — the same
  objection that kept the platform string coarse, and applying it to one field and not the other
  was the inconsistency.

`numpy-blas` carries the migration on its own: `blas` under conda, `accelerate` after the move.

**When to add it.** If someone needs to distinguish *which* implementation a conda `blas` resolves
to — the one thing `show_config` cannot answer — then `threadpoolctl` becomes a dev-group
dependency, the field records implementation and version but **not** the thread count, and its null
needs `threadpoolctl`'s own version recorded beside it so "not installed" reads differently from
"installed and silent". That last point is not free: the biconditional does not hold as stated,
because a vendored or `PYTHONPATH`-mounted copy has no dist-info while `threadpool_info()` still
answers.

Prior art, for whoever picks this up: scikit-learn hard-depends on threadpoolctl and dumps
`threadpool_info()` from `sklearn/utils/_show_versions.py`; pandas and polars record the platform
and no BLAS at all.
